src/Custom/Microsoft.Extensions.AI/OpenAIChatClient.cs

```diff
@@ -0,0 +1,760 @@
+            counts.Add($"{InputDetails}.{nameof(ChatInputTokenUsageDetails.CachedTokenCount)}", inputDetails.CachedTokenCount);
+        }
+
+        if (tokenUsage.OutputTokenDetails is ChatOutputTokenUsageDetails outputDetails)
+        {
+            const string OutputDetails = nameof(ChatTokenUsage.OutputTokenDetails);
+            counts.Add($"{OutputDetails}.{nameof(ChatOutputTokenUsageDetails.ReasoningTokenCount)}", outputDetails.ReasoningTokenCount);
+            counts.Add($"{OutputDetails}.{nameof(ChatOutputTokenUsageDetails.AudioTokenCount)}", outputDetails.AudioTokenCount);
+            counts.Add($"{OutputDetails}.{nameof(ChatOutputTokenUsageDetails.AcceptedPredictionTokenCount)}", outputDetails.AcceptedPredictionTokenCount);
+            counts.Add($"{OutputDetails}.{nameof(ChatOutputTokenUsageDetails.RejectedPredictionTokenCount)}", outputDetails.RejectedPredictionTokenCount);
+        }
+
+        return destination;
+    }
+
+    /// <summary>Converts an OpenAI role to an Extensions role.</summary>
+    private static ChatRole FromOpenAIChatRole(ChatMessageRole role) =>
+        role switch
+        {
+            ChatMessageRole.System => ChatRole.System,
+            ChatMessageRole.User => ChatRole.User,
+            ChatMessageRole.Assistant => ChatRole.Assistant,
+            ChatMessageRole.Tool => ChatRole.Tool,
+            ChatMessageRole.Developer => ChatRoleDeveloper,
+            _ => new ChatRole(role.ToString()),
+        };
+
+    /// <summary>Creates an <see cref="AIContent"/> from a <see cref="ChatMessageContentPart"/>.</summary>
+    /// <param name="contentPart">The content part to convert into a content.</param>
+    /// <returns>The constructed <see cref="AIContent"/>, or <see langword="null"/> if the content part could not be converted.</returns>
+    private static AIContent? ToAIContent(ChatMessageContentPart contentPart)
+    {
+        AIContent? aiContent = null;
+
+        if (contentPart.Kind == ChatMessageContentPartKind.Text)
+        {
+            aiContent = new TextContent(contentPart.Text);
+        }
+        else if (contentPart.Kind == ChatMessageContentPartKind.Image)
+        {
+            aiContent =
+                contentPart.ImageUri is not null ? new UriContent(contentPart.ImageUri, "image/*") :
+                contentPart.ImageBytes is not null ? new DataContent(contentPart.ImageBytes.ToMemory(), contentPart.ImageBytesMediaType) :
+                null;
+
+            if (aiContent is not null && contentPart.ImageDetailLevel?.ToString() is string detail)
+            {
+                (aiContent.AdditionalProperties ??= [])[nameof(contentPart.ImageDetailLevel)] = detail;
+            }
+        }
+
+        if (aiContent is not null)
+        {
+            if (contentPart.Refusal is string refusal)
+            {
+                (aiContent.AdditionalProperties ??= [])[nameof(contentPart.Refusal)] = refusal;
+            }
```

Comment:
I think the two uses of `AdditionalProperties` in this method can be removed since they can be read from `RawRepresentation` instead.